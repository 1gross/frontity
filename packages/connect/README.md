## Description
The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.

You can find more information about it at [Frontity docs](https://docs.frontity.org/javascript-and-and-react/react-basic#state-manager-frontity-connect).
