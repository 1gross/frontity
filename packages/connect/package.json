{
  "name": "@frontity/connect",
  "version": "0.1.0",
  "description": "A state manager that's easy to use and hackable at the same time.",
  "keywords": [
    "frontity",
    "connect"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "jest",
    "test": "jest --watch",
    "test:inspect": "node --inspect ./node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "recursive-iterator": "^3.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@types/jest": "^24.0.11",
    "babel-jest": "^24.8.0",
    "jest": "^24.7.1",
    "ts-jest": "^24.0.2",
    "typescript": "^3.4.2"
  }
}
