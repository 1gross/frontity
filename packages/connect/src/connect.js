import { createContext, useContext, createElement } from "react";
import { view } from "@risingstack/react-easy-state";

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext();

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * Connects a React component with the Frontity connect store.
 *
 * @param Component The React component to be connected
 * @returns The same Component, but with the store included in the props.
 */
export const connect = (Component) => {
  const isStatelessComp = !(
    Component.prototype && Component.prototype.isReactComponent
  );

  if (isStatelessComp) {
    return view((props) => {
      const store = useContext(context);
      return Component({ ...props, ...store });
    });
  } else {
    class ConnectedComponent extends Component {
      static contextType = context;

      render() {
        return createElement(Component, {
          ...this.props,
          ...this.context,
        });
      }
    }

    return view(ConnectedComponent);
  }
};
