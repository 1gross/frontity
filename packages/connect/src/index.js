export { observe, unobserve } from "./observer";
export { observable, isObservable, raw, clone } from "./observable";
export { createStore, getSnapshot } from "./create-store";
export { connect as default, Provider, useConnect } from "./connect";
export { batch } from "./scheduler";
