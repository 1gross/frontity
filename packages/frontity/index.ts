import create, { events } from "./src/commands/create";

events.on("create", console.log);
create({ name: "random-name" });
