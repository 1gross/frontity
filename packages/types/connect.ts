import Package from "./package";
import { ResolveState, Omit } from "./utils";

// Resolves the actions for things like Action and Derived.
export type ResolveActions<Actions extends Package["state"]> = {
  [P in keyof Actions]: Actions[P] extends (
    ...a: any
  ) => (arg: infer Arg) => void // Turns "state => args => {}" into "args => {}"
    ? (arg: Arg) => void // Turns "state => {}" into "() => {}"
    : Actions[P] extends (state: Package["state"]) => any
    ? () => void
    : ResolveActions<Actions[P]>
};

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export default Connect;
