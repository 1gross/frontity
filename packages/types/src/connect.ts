import Package from "./package";
import { ResolveState, Omit, ResolveActions } from "./utils";

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export type ExternalProps<T extends Package> = Omit<
  T,
  "state" | "actions" | "libraries" | "roots" | "fills"
>;

export default Connect;
