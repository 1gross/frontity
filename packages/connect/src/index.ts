import { Store, State, ProxifyOptions, Context, Actions } from "../types";

export const proxifyState = <StoreType extends Store>(
  rawStore: StoreType,
  context?: Context,
  options?: ProxifyOptions
): State<StoreType> => {
  return rawStore.state as State<StoreType>; // Overwrite type until it's implemented.
};

export const proxifyActions = <StoreType extends Store>(
  rawStore: StoreType,
  context?: Context,
  options?: ProxifyOptions
): Actions<StoreType> => {
  return rawStore.actions as Actions<StoreType>; // Overwrite type until it's implemented.
};
