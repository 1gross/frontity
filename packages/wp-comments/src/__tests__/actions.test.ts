import * as frontity from "frontity";
import { createStore } from "@frontity/connect";
import clone from "clone-deep";
import wpComments from "..";
import WpComments from "../../types";

const fetch = jest.spyOn(frontity, "fetch").mockResolvedValue(undefined);
const warn = jest.spyOn(frontity, "warn");

describe("WP Comments > create", () => {
  beforeEach(() => {
    fetch.mockClear();
    warn.mockClear();
  });

  test("should send the correct call to WordPress", async () => {
    // Init mocked store with empty state.source.
    const store = createStore<WpComments>(
      Object.assign(clone(wpComments), { state: { source: {} } })
    );

    // Set a test api that should be used to generate the comment endpoint.
    store.state.source.api = "https://test.frontity.org/wp-json/";

    // Send a comment.
    await store.actions.comments.create({
      comment: "This is a comment example. Hi!",
      author: "Frontibotito",
      email: "frontibotito@frontity.com",
      url: "https://frontity.org",
      postId: 60,
    });

    // Check the comment was sent.
    expect(fetch.mock.calls).toMatchSnapshot();

    // Check the request body is correct.
    const { body } = fetch.mock.calls[0][1];
    expect(body.toString()).toMatchSnapshot();
  });

  test("should show a warning message if the source is a WP.com site", () => {
    // Init mocked store with empty state.source.
    const store = createStore<WpComments>(
      Object.assign(clone(wpComments), { state: { source: {} } })
    );

    // Mock that `source.api` points to a WordPress.com site.
    store.state.source.isWpCom = true;

    // Send a comment.
    store.actions.comments.create({
      comment: "This is a comment example. Hi!",
      author: "Frontibotito",
      email: "frontibotito@frontity.com",
      url: "https://frontity.org",
      postId: 60,
    });

    // Check the comment was not sent.
    expect(fetch).not.toHaveBeenCalled();

    // Check that a warning message was shown instead.
    expect(warn.mock.calls).toMatchSnapshot();
  });
});
