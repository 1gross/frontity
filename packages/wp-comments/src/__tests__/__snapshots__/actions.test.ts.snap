// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WP Comments > create should send the correct call to WordPress 1`] = `
Array [
  Array [
    "https://test.frontity.org/wp-comments-post.php",
    Object {
      "body": URLSearchParams {},
      "headers": Object {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`WP Comments > create should send the correct call to WordPress 2`] = `"comment=This+is+a+comment+example.+Hi%21&author=Frontibotito&email=frontibotito%40frontity.com&url=https%3A%2F%2Ffrontity.org&comment_post_ID=60&comment_parent=0"`;

exports[`WP Comments > create should show a warning message if the source is a WP.com site 1`] = `
Array [
  Array [
    "Sending comments to a WordPress.com site is not supported yet.",
  ],
]
`;
